src/uct/tcp/tcp_ep.c

```diff
@@ -467,51 +466,25 @@ void uct_tcp_ep_mod_events(uct_tcp_ep_t *ep, int add, int remove)
     }
 }
 
-static inline void uct_tcp_ep_unacked_puts_advance(uct_tcp_ep_ctx_t *ctx,
-                                                   unsigned new_unacked)
-{
-    ucs_assert(ctx->unacked_put_sn >= ctx->acked_put_sn);
-    ctx->unacked_put_sn += new_unacked;
-}
-
-static inline void uct_tcp_ep_acked_puts_advance(uct_tcp_ep_ctx_t *ctx,
-                                                 unsigned new_acked)
-{
-    ctx->acked_put_sn += new_acked;
-    ucs_assert(ctx->unacked_put_sn >= ctx->acked_put_sn);
-}
-
-static inline unsigned uct_tcp_ep_unacked_puts_cnt(uct_tcp_ep_ctx_t *ctx)
-{
-    return ctx->unacked_put_sn - ctx->acked_put_sn;
-}
-
-static inline void uct_tcp_ep_commit_put_req(uct_tcp_ep_t *ep)
-{
-    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
-                                            uct_tcp_iface_t);
-
-    uct_tcp_ep_unacked_puts_advance(&ep->tx, 1);
-
-    uct_tcp_iface_outstanding_inc(iface);
-}
-
-static inline void uct_tcp_ep_commit_put_ack(uct_tcp_ep_t *ep,
-                                             unsigned acked_put_cnt)
+static inline void uct_tcp_ep_handle_put_ack(uct_tcp_ep_t *ep,
+                                             unsigned acked_put_sn)
 {
     uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
                                             uct_tcp_iface_t);
     uct_tcp_ep_put_completion_t *put_comp;
 
-    uct_tcp_ep_acked_puts_advance(&ep->tx, acked_put_cnt);
-
     uct_tcp_iface_outstanding_dec(iface);
 
+    if (acked_put_sn == ep->tx.put_sn) {
+        /* Since there are no other PUT operations in-flight, can remove flag */
+        ep->ctx_caps &= ~UCS_BIT(UCT_TCP_EP_CTX_TYPE_PUT_TX_WAITING_ACK);
+    }
+
     while (!ucs_queue_is_empty(&ep->put_comp_q)) {
         put_comp = ucs_queue_head_elem_non_empty(&ep->put_comp_q,
                                                  uct_tcp_ep_put_completion_t,
                                                  elem);
-        if (put_comp->wait_put_ack_sn > ep->tx.acked_put_sn) {
+        if (put_comp->wait_put_ack_sn > acked_put_sn) {
```

Comment:
must use CIRCULAR_COMPARE
define sn types as uint32_t instead of unsifgned

Reply:
good catch!
than you
done